pkg/scanners/trivy/trivy.go

```diff
@@ -0,0 +1,397 @@
+		// note: this function has side effects and will modify the map supplied as the first argument
+		err := parseCommaSeparatedOptions(oSet.m, oSet.input)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			os.Exit(generalErr)
+		}
+	}
+
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(generalErr)
+	}
+
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(generalErr)
+	}
+
+	result := eraserv1alpha1.ImageCollector{}
+
+	err = clientset.RESTClient().Get().
+		AbsPath(apiPath).
+		Resource(resourceName).
+		Name(*collectorCRName).
+		Do(context.Background()).
+		Into(&result)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(generalErr)
+	}
+
+	// deduplicate, which may not be necessary
+	scanList := make(map[string]string)
+	for _, img := range result.Spec.Images {
+		scanList[img.Digest] = img.Name
+	}
+
+	err = downloadAndInitDB(*cacheDir)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(generalErr)
+	}
+
+	vulnTypeList := trueMapKeys(vulnTypeMap)
+	securityCheckList := trueMapKeys(securityCheckMap)
+
+	scanConfig, err := setupScanner(*cacheDir, vulnTypeList, securityCheckList)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(generalErr)
+	}
+
+	resultClient := initializeResultClient()
+
+	imgChan := make(chan string)
+	var wg sync.WaitGroup
+
+	for _, imageName := range scanList {
+		wg.Add(1)
+
+		go func(imageRef string) {
+			fmt.Printf("scanning: %s\n", imageRef)
+			dockerImage, cleanup, err := fanalImage.NewDockerImage(ctx, imageRef, scanConfig.dockerOptions)
+			if err != nil {
+				fmt.Fprintln(os.Stderr, err)
+				os.Exit(generalErr)
+			}
+			defer cleanup()
+
+			artifactToScan, err := artifactImage.NewArtifact(dockerImage, scanConfig.fscache, artifact.Option{}, config.ScannerOption{})
+			if err != nil {
+				fmt.Fprintln(os.Stderr, err)
+				os.Exit(generalErr)
+			}
+
+			scanner := scanner.NewScanner(scanConfig.localScanner, artifactToScan)
+
+			report, err := scanner.ScanArtifact(ctx, scanConfig.scanOptions)
+			if err != nil {
+				fmt.Fprintln(os.Stderr, err)
+				os.Exit(generalErr)
+			}
+
+		outer:
+			for i := range report.Results {
+				resultClient.FillVulnerabilityInfo(report.Results[i].Vulnerabilities, report.Results[i].Type)
+
+				for j := range report.Results[i].Vulnerabilities {
+					if *ignoreUnfixed && report.Results[i].Vulnerabilities[j].FixedVersion == "" {
+						continue
+					}
+
+					if report.Results[i].Vulnerabilities[j].Severity == "" {
+						report.Results[i].Vulnerabilities[j].Severity = severityUnknown
+					}
+
+					if severityMap[report.Results[i].Vulnerabilities[j].Severity] {
+						imgChan <- report.ArtifactName
+						break outer
+					}
+				}
+			}
+
+			wg.Done()
+		}(imageName)
+	}
+
+	go func() {
+		wg.Wait()
+		close(imgChan)
+	}()
+
+	vulnerableImages := make([]eraserv1alpha1.Image, 0, len(scanList))
```

Comment:
should we separate out the reporting back logic so it can be reused later?

Reply:
@sozercan Would you mind elaborating? I'm not sure I follow what you mean

Reply:
each provider processes collector results and outputs a list, right? say, if we have a different scanner tomorrow and can we have share outputting of the scanner results?